### PR TITLE
Performance: Refactor resolving SpanContext for Throwable

### PR DESCRIPTION
**Pull request by maciejwalkowiak**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring

## :scroll: Description

Refactor resolving SpanContext for Throwable

## :bulb: Motivation and Context

Previous implementation would fail once we switch to sending individual spans. Here we keep an association between throwable and SpanContext on the `Hub` which is independent from transaction or span being still in memory. 

## :green_heart: How did you test it?

Integration tests cover it.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: sentry/src/main/java/io/sentry/Span.java [hunk 1]
```diff
@@ -52,6 +56,9 @@ public SentryTraceHeader toSentryTrace() {
   @Override
   public void finish() {
     this.timestamp = DateUtils.getCurrentDateTime();
+    if (this.throwable != null) {
+      hub.setSpanContext(this.throwable, this);
+    }
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/Span.java:61]: just a `nit`, we don't need to use `this.` if there's no class/parameters shadowing

**Comment by maciejwalkowiak** [comment]
@bruno-garcia @marandaneto please take a look.

**Comment by maciejwalkowiak** [comment]
As far as I can tell covecov reports are broken. Since we have a check for code coverage in gradle build already, could we turn it off?

File: sentry/src/main/java/io/sentry/IHub.java [hunk 2]
```diff
@@ -333,6 +333,27 @@ default SentryTransaction startTransaction(final @NotNull String name) {
...
+  /**
+   * Gets the span context for the span that was active while the throwable given by parameter was
+   * thrown.
+   *
+   * @param throwable - the throwable
+   * @return span context or {@code null} if no corresponding span context found.
+   */
+  @ApiStatus.Internal
+  @Nullable
+  SpanContext getSpanContext(Throwable throwable);
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/IHub.java:355]:
```suggestion
  SpanContext getSpanContext(@NotNull Throwable throwable);
```

File: sentry/src/main/java/io/sentry/NoOpHub.java [hunk 3]
```diff
@@ -133,6 +133,14 @@ public SentryTransaction startTransaction(
     return new SentryTraceHeader(SentryId.EMPTY_ID, SpanId.EMPTY_ID, true);
   }
 
+  @Override
+  public void setSpanContext(@NotNull Throwable throwable, @NotNull SpanContext spanContext) {}
+
+  @Override
+  public @Nullable SpanContext getSpanContext(Throwable throwable) {
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/NoOpHub.java:140]:
```suggestion
  public @Nullable SpanContext getSpanContext(@NotNull Throwable throwable) {
```

**Review by bruno-garcia — APPROVED** [review]
🚀

**Pull request merged**